### PR TITLE
[Issue: 3161][Go client] Return Message ID for send

### LINKED ISSUE #3161: Message ID of sent message should be returned in Go client producer
#### Expected behavior

Go client library returns the message ID of a message that has been commited to disk. Just like the Java client currently does. Both for async and sync send.

#### Actual behavior

There is no way of getting the message ID of a sent message in the current implementation of the Go client as far as I can tell.

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <ranxiaolong716@gmail.com>

Fixes #3161 #3180 #4863

### Motivation

The `Send(context.Context, ProducerMessage) error` should return `MessageID`.

```
 SendAsync(context.Context, ProducerMessage, func(ProducerMessage, error))
```

In here, we shouldn't return `ProducerMessage`, because the value is nil.

### CODE REVIEW

**Comment by wolfstudy** [comment]
@sijie @jiazhai @merlimat PTAL

**Comment by sijie** [comment]
I think the change looks good to me, and it makes the API of C++/Go/Python consistent with Java. However since this change change the callback signatures, it would be great to see the opinions from cpp client users. 

@merlimat @rdhabalia @massakam can you guys please take a look at this change?

**Comment by sijie** [comment]
ping @merlimat @rdhabalia @massakam ?

**Comment by wolfstudy** [comment]
hello @merlimat @rdhabalia @massakam PTAL thanks.

**Comment by wolfstudy** [comment]
Png @merlimat @rdhabalia @massakam If time is convenient, please review the pull request, thanks.

**Comment by wolfstudy** [comment]
Ping @merlimat @rdhabalia @massakam If time is convenient, please review the pull request, thanks again.

**Comment by wolfstudy** [comment]
Ping @merlimat @rdhabalia @massakam If time is convenient, please review the pull request, thanks again.

**Comment by sijie** [comment]
Ping @merlimat @rdhabalia @massakam

**Comment by massakam** [comment]
Sorry for the late response. This change looks good to me too.

File: pulsar-client-go/pulsar/producer.go [hunk 1]
```diff
@@ -166,13 +166,26 @@ type Producer interface {
...
 	// Example:
 	// producer.Send(ctx, pulsar.ProducerMessage{ Payload: myPayload })
+	// @Deprecated
 	Send(context.Context, ProducerMessage) error
 
+	// Send a message
+	// This call will be blocking until is successfully acknowledged by the Pulsar broker.
+	// Example:
+	// msgID, err := producer.SendWithMsgID(ctx, pulsar.ProducerMessage{ Payload: myPayload })
+	SendWithMsgID(context.Context, ProducerMessage) (MessageID, error)
```
Comment (massakam) [inline pulsar-client-go/pulsar/producer.go]: From this method name, I get the impression that "Message ID is sent to broker with something" or "Something is sent to broker using message ID". What about a name like `SendAndGetMsgId`?

**Comment by wolfstudy** [comment]
Sorry for the late response. @massakam PTAL again thanks.

**Comment by wolfstudy** [comment]
ping @massakam @merlimat PTAL

**Comment by wolfstudy** [comment]
@massakam PTAL again thanks

**Comment by wolfstudy** [comment]
ping @massakam PTAL

**Comment by wolfstudy** [comment]
ping @massakam PTAL again thanks

**Comment by massakam** [comment]
rerun java8 tests
rerun cpp tests

**Comment by massakam** [comment]
rerun java8 tests

**Comment by massakam** [comment]
rerun java8 tests

**Comment by massakam** [comment]
rerun java8 tests

**Comment by massakam** [comment]
rerun java8 tests

**Comment by massakam** [comment]
rerun java8 tests

**Comment by massakam** [comment]
@merlimat What do you think about this change?

**Comment by wolfstudy** [comment]
ping @merlimat PTAL

**Comment by vicaya** [comment]
Log here for posterity, pre 2.4 python client e.g. 2.3.2 would segfault if people use send_async and try to inspect msg.top_name() etc in the call back. 2.4+ python client would correctly complain about attribute errors on MessageId object.

**Pull request merged**